CARDS-1780: Configure/Implement data import from Clarity

OrdiNeu opened the pull request:
This PR involves the clarity importer, and includes sample testing data

## To test (via Docker):
1. Build with `-Pdocker` and navigate to the compose-cluster folder.
2. Clean up any old instance of CARDS with `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
3. Start up Docker with `python3 generate_compose_yaml.py --mssql --cards_project cards4prems --oak_filesystem --dev_docker_image --composum --adminer` (of note: `--adminer` to setup adminer, `--cards_project cards4prems` for the prems project, `--mssql` to setup the MS-SQL container, and `--custom_env_file mssql/mssql.env` to set it up with the environment variables necessary.)
  - As an aside, the `compose-cluster/mssql/mssql.env` file is there to show which environment variables are important to setup CARDs' clarity importer.
4. `docker-compose build && docker-compose up -d` to start up the server 
5. Navigate to http://localhost:1435/ to access adminer. Login with the following credentials:
- System: `MS SQL (beta)`
- Server: `mssql`
- Username: `sa`
- Password: `testPassword_`
- Database: `master`
6. On the left bar, click Import.
7. On the next screen, upload the file `cards/compose-cluster/mssql/importTestSql.sql` like so:
![image](https://user-images.githubusercontent.com/4656440/200427919-a8fc69a9-cf6e-494c-80d1-6d98ec9c1117.png)
- It should say "4 queries executed OK." if it was successful.
- You can add more rows if you want to test other scenarios
- I've also included a tool inside `compose-cluster/mssql/` called `generate_test_sql.py`. Call it with `python3 generate_test_sql.py test.sql -n 100` to generate 100 entries
9. Login to cards with the admin account
10. Navigate to `http://localhost:8080/Subjects.importClarity`
- You should receive a white page with `{"status":"success","message":"Data import started"}`
11. The patients and forms should be imported momentarily. I have noticed a delay of a minute or five when importing more than 1000 forms.
12. When finished, `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`

OrdiNeu committed:
CARDS-1780: Configure/Implement data import from Clarity

OrdiNeu committed:
CARDS-1780: Made the mssql driver accessible

OrdiNeu committed:
CARDS-1780: Reuse subjects that exist, check in forms after creating
them, code cleanup

OrdiNeu committed:
CARDS-1780: Add testing

OrdiNeu committed:
CARDS-1780: Add testing environment

OrdiNeu committed:
CARDS-1780: Fix up the SQL importer to be fault-safe, and fix the
default password on the importer

OrdiNeu committed:
CARDS_1780: Code cleanup

IntegralProgrammer commented on compose-cluster/mssql/importTestSql.sql:
```diff
@@ -0,0 +1,53 @@
+INSERT INTO [path].[CL_EP_IP_EMAIL_CONSENT_IN_LAST_7_DAYS]
+    (PAT_MRN, EMAIL_ADDRESS, ENTRY_TIME, DISCH_DEPT_NAME, EMAIL_CONSENT_YN, LoadTime)
+    VALUES
+    (1234568, 'test@test.com', '2022-09-21T11:16:20.136Z', 'TW-EMERGENCY', 'No', CAST(GETDATE() AS DATE))
```
Would we want a script to generate this? Perhaps one that can be parameterized to select names and times at random such as what can be done with https://github.com/data-team-uhn/cards/blob/dev/Utilities/Development/mock_torch.js?

IntegralProgrammer commented on prems-resources/clarity-import/src/main/java/io/uhndata/cards/internal/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,375 @@
+        String connectionUrl =
+            "jdbc:sqlserver://" + System.getenv("SQL_SERVER") + ";"
+            + "user=" + System.getenv("SQL_USERNAME") + ";"
+            + "password=" + System.getenv("SQL_PASSWORD") + ";"
```
We should use more _specifically_ named environment variables. Perhaps in the future we might need to connect CARDS to other SQL servers. How about `CLARITY_SQL_SERVER`, `CLARITY_SQL_USERNAME`, and `CLARITY_SQL_PASSWORD`?

IntegralProgrammer commented on prems-resources/clarity-import/src/main/java/io/uhndata/cards/internal/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,375 @@
+            "jdbc:sqlserver://" + System.getenv("SQL_SERVER") + ";"
+            + "user=" + System.getenv("SQL_USERNAME") + ";"
+            + "password=" + System.getenv("SQL_PASSWORD") + ";"
+            + "encrypt=false;";
```
Database connection encryption should be configurable. I would say that it should be _enabled_ by default (thus making the most secure behavior as the default) with an option to disable it.

OrdiNeu replied:
Though I can set it to `true` by default, the connection will fail into the local mssql server fails due to no certifications on localhost.

IntegralProgrammer replied:
It should still default to `true` and when `generate_compose_yaml.py` is called with the `--mssql` flag, the `CLARITY_SQL_ENCRYPT` environment variable for the `cardsinitial` container should be set to `false`.

We do a similar thing with `--smtps_test_container` at https://github.com/data-team-uhn/cards/blob/53fecd129e092b8bbd8d763c6a5bcd59275b49e9/compose-cluster/generate_compose_yaml.py#L517-L519 By setting the `SMTPS_LOCAL_TEST_CONTAINER` environment variable to `true`, `docker_entry.sh` is instructed to instruct CARDS to not check the SMTPS server's identity. See https://github.com/data-team-uhn/cards/blob/53fecd129e092b8bbd8d763c6a5bcd59275b49e9/distribution/docker_entry.sh#L183-L188

IntegralProgrammer commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -699,6 +700,19 @@ def getCardsJavaMemoryLimitMB():
+if args.mssql:
+  print("Configuring service: ms-sql")  
+  yaml_obj['services']['mssql'] = {
+    'image': 'mcr.microsoft.com/mssql/server',
```
We should lock this image to a certain _tag_ (eg. `2022-latest`) if possible.

IntegralProgrammer commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -699,6 +700,19 @@ def getCardsJavaMemoryLimitMB():
+        'aliases': ['mssql']
+      }
+    },
+    'ports': ['127.0.0.1:1433:1433'],
```
For most testing/development purposes, we won't actually need to connect to MS SQL from the host, although it could be a nice option to have. For that reason, I suggest having an optional `--expose_mssql <PORT NUMBER>` argument that, when specified, forwards a _localhost_ port to _TCP:1433_ on the `mssql` container. For example, `python3 generate_compose_yaml.py --expose_mssql 1433 ...` would make the MS SQL database accessible at `localhost:1433` from the host OS.

IntegralProgrammer commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -699,6 +700,19 @@ def getCardsJavaMemoryLimitMB():
 
+if args.mssql:
+  print("Configuring service: ms-sql")  
+  yaml_obj['services']['mssql'] = {
+    'image': 'mcr.microsoft.com/mssql/server',
+    'networks': {
+      'internalnetwork': {
+        'aliases': ['mssql']
+      }
+    },
+    'ports': ['127.0.0.1:1433:1433'],
+    'environment': ['ACCEPT_EULA=Y', 'MSSQL_SA_PASSWORD=testPassword_']
+  }
```
Although this does create the correct structure in the generated YAML file, I think, for the sake of consistency, it would be better to follow the same code style as that which is used to configure the  _Adminer_ or _S3_ container.

IntegralProgrammer commented on compose-cluster/mssql/mssql.env:
```diff
@@ -0,0 +1,20 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+SQL_SERVER="mssql:1433"
+SQL_USERNAME="sa"
+SQL_PASSWORD="testPassword_"
```
Instead of having this file and needing to specify `--custom_env_file`, it would be better if these environment variables were configured automatically for the `cardsinitial` container whenever the `--mssql` flag is specified.

OrdiNeu replied:
For multiple variables that need to be configured differently between testing and prod, I think a custom env file would be better. When putting this up on prod, we'll need to configure these parameters to point at the actual MS-SQL server, and won't be using the `-mssql` flag.

IntegralProgrammer replied:
> When putting this up on prod, we'll need to configure these parameters to point at the actual MS-SQL server, and won't be using the `-mssql` flag.

That's correct, we will not be using the `--mssql` flag as the _only_ purpose of that flag is to run a MS-SQL Docker container for testing/development purposes. If we're running a MS-SQL Docker container for testing/development purposes, then wouldn't it make sense to automatically point the CARDS container to it (otherwise, why would we be starting a test MS-SQL Docker container)?

For _production_ environments, we won't specify the `--mssql` flag so no `CLARITY_SQL*` environment variables would be automatically configured - either we would do this through a `--custom_env_file` or by manually editing the `docker-compose.yml` file before starting up the Docker environment.

OrdiNeu replied:
Right but, we should definitely have this file or documentation somewhere as a reference for which variables need to be configured. Otherwise people would have to look at the code to figure out how to set up prod.

OrdiNeu replied:
Implemented in [80b88aa](https://github.com/data-team-uhn/cards/pull/1204/commits/80b88aa45130ee208137ea83e3a9ad3a9286a138) :heavy_check_mark:

OrdiNeu committed:
CARDS-1780: Apply code review comments by @IntegralProgrammer

IntegralProgrammer commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -54,6 +54,8 @@
 argparser.add_argument('--external_mongo_uri', help='URI of the external MongoDB instance. Only valid if --external_mongo is specified.')
 argparser.add_argument('--external_mongo_dbname', help='Database name of the external MongoDB instance. Only valid if --external_mongo is specified.')
+argparser.add_argument('--mssql', help='Start up a MS-SQL instance with test data', action='store_true')
+argparser.add_argument('--expose_mssql', help='If --mssql is specified, forward the SQL service to the specified port [default: 1433]', default=1433, type=int)
```
By setting a default value here, `args.mssql` will be set to the default value (`1433`, in this case) whenever `--expose_mssql` is not specified - which is unexpected behavior as one would not expect MS-SQL to be exposed to the host if `--expose_mssql` is not specified.

IntegralProgrammer commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -699,6 +701,17 @@ def getCardsJavaMemoryLimitMB():
+  yaml_obj['services']['mssql']['networks']['internalnetwork']['aliases'] = ['mssql']
+  yaml_obj['services']['mssql']['environment'] = ['ACCEPT_EULA=Y', 'MSSQL_SA_PASSWORD=testPassword_']
+  if args.expose_mssql:
+    yaml_obj['services']['mssql']['ports'] = ['127.0.0.1:1433:1433']
```
```suggestion
    yaml_obj['services']['mssql']['ports'] = ['127.0.0.1:{}:1433'.format(args.expose_mssql)]
```

IntegralProgrammer commented on compose-cluster/mssql/generateTestSQL.py:
```diff
@@ -0,0 +1,90 @@
+
+argparser = argparse.ArgumentParser()
+argparser.add_argument('file', help='Output file name', type=argparse.FileType('w'))
+argparser.add_argument('-n', help='Number of arguments [default: 3]', default=3, type=int)
```
Change the `help` parameter to something more descriptive - `Number of patient discharges to be generated`?

IntegralProgrammer commented on compose-cluster/mssql/generateTestSQL.py:
```diff
@@ -0,0 +1,90 @@
+    email = 'test' + str(mrn) + '@test.com'
+    entry_time = datetime.date.today() - datetime.timedelta(seconds=random.randint(0, 7 * 24 * 60 * 60))
+    entry_time_str = entry_time.strftime('%Y-%m-%d %H:%M:%S')
+    disch_dept_name = ('TG-EMERGENCY', 'TW-EMERGENCY')[random.randint(0, 1)]
+    email_consent_yn = ('Yes', 'No')[random.randint(0, 1)]
```
Using `random.choice` results in slightly cleaner code.

IntegralProgrammer commented on compose-cluster/mssql/generateTestSQL.py:
```diff
@@ -0,0 +1,90 @@
+for i in range(args.n):
+    mrn = random.randint(0, 9999999)
+    email = 'test' + str(mrn) + '@test.com'
+    entry_time = datetime.date.today() - datetime.timedelta(seconds=random.randint(0, 7 * 24 * 60 * 60))
```
This should be controllable through the CLI arguments. How about a `--base_date` parameter that defaults to today's date and a `--time_spread_seconds` parameter that defaults to `7 * 24 * 60 * 60`?

IntegralProgrammer commented on compose-cluster/mssql/mssql.env:
```diff
@@ -0,0 +1,21 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+CLARITY_SQL_SERVER="mssql:1433"
+CLARITY_SQL_USERNAME="sa"
+CLARITY_SQL_PASSWORD="testPassword_"
+CLARITY_SQL_ENCRYPT="false"
```
Please see https://github.com/data-team-uhn/cards/pull/1204#discussion_r1018133417. If you agree that these environment variables should be automatically set whenever the `--mssql` flag is specified, then this file is no longer needed.

OrdiNeu replied:
We need to document these variables somewhere. If not here, then maybe in a `.md`? What do you think?

OrdiNeu committed:
CARDS-1780: Move mssql configuration into the dockerfile

OrdiNeu committed:
CARDS-1780: Apply code review comments by @IntegralProgrammer

OrdiNeu committed:
CARDS-1780: Rename generate_test_sql.py to match the other python
scripts, and add new arguments

sdumitriu commented on compose-cluster/generate_compose_yaml.py:
```diff
@@ -54,6 +54,8 @@
 argparser.add_argument('--external_mongo_uri', help='URI of the external MongoDB instance. Only valid if --external_mongo is specified.')
 argparser.add_argument('--external_mongo_dbname', help='Database name of the external MongoDB instance. Only valid if --external_mongo is specified.')
+argparser.add_argument('--mssql', help='Start up a MS-SQL instance with test data', action='store_true')
+argparser.add_argument('--expose_mssql', help='If --mssql is specified, forward the SQL service to the specified port', type=int)
```
Should this default to `1433`?

IntegralProgrammer replied:
Addressed in https://github.com/data-team-uhn/cards/pull/1204/commits/a3ede5fa93b98b4f3f5d8967dd0423cd9e0c642f.

- If `--expose_mssql` is _not_ specified, then MS SQL will _not_ be available at any localhost port.
- If `--expose_mssql` is specified, then MS SQL will be available at `127.0.0.1:1433`.
- If `--expose_mssql [PORT]` is specified, then MS SQL will be available at `127.0.0.1:[PORT]` (eg. `--expose_mssql 5555` will cause MS SQL to be available at `127.0.0.1:5555`).

sdumitriu commented on distribution/src/main/features/core/cards.json:
```diff
@@ -72,6 +72,10 @@
             "start-order":"25"
+        },
+        {
+            "id":"io.uhndata.cards:cards-clarityimport:${project.version}",
```
I don't like that the core distribution depends on a submodule of cards4prems.

sashaandjic commented:
@OrdiNeu : I'm not sure if I missed something. I followed all the steps and all looked good, however I do not see any forms being imported
![image](https://user-images.githubusercontent.com/23243244/203423542-e028e4e6-8291-4cb4-9a5c-292cf9775119.png)
![image](https://user-images.githubusercontent.com/23243244/203423717-d34ae5e9-b33a-4587-bd35-172a5ca4f235.png)

sashaandjic reviewed (approved):
tested

I was able to import 103 patients from 2 different sources as per instructions provided in the PR. All good!

OrdiNeu committed:
CARDS-1780: Move environment variable documentation from a .env file to
the environment.md file

marta- reviewed (commented):
IMO the java package should be moved into a core module. I suggest the name `clarity-integration`.

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactor code moving it into the clarity-integration core CARDS module

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Clean up feature.json files related to the Clarity data importer

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Load the Clarity data importer module if start_cards.sh is called with --clarity or with --project cards4prems

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Docker support: Load the Clarity data importer module if generate_compose_yaml.py is called with --clarity or with --cards_project cards4prems

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix the clarity-import service user permissions issues

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix issue where multiple CARDS Patient subjects with the same identifier were being created

TODO: Fix issue where multiple Patient information forms for the same Patient subject are being created

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix issue where multiple Patient information forms for the same Patient subject are being created

TODO: Ensure that this bug does not happen when the same PAT_MRN is imported on subsequent Clarity imports

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Update Patient information forms on each /Subjects.importClarity operation with the response data from Clarity

TODO: Code clean up and refactoring

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix bug where import of ENTRY_TIME column from the Clarity SQL database was ignoring the time component and only keeping the date component (eg. '2022-11-30 14:30:00' would be stored as '2022-11-30 00:00:00')

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - remove extra redundant call to props.put(ClarityImportTask.VALUE_PROP, calendar)

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - clean up commenting

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - remove debug logging

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - clean up commenting

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - remove debug logging

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - remove debug logging

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix Clarity importer so that boolean yes/no values are stored as long values in the JCR as this is the way that this type of data is handled when it is entered through the web UI

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - assign the 'private' access modifier to everything in the ClarityImportTask class that should be private

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code refactoring - create the generateAnswerNodeProperties function

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code refactoring - create the getUpdatableForm function

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup - remove the SuppressWarnings directive now that the code has been cleaned up and these warnings are no longer generated

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code refactoring - introduce the formAlreadyCreated and getSubjectForResult methods

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Commenting cleanup

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactored code by renaming newSubject variable to subject and newForm variable to form as these Subjects and Forms are not necessairly new - we might simply be updating previously existing Subjects and Forms

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactored code by introducing the markFormAsCreated method

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix commenting

veronikaslc commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportEndpoint.java:
```diff
@@ -0,0 +1,100 @@
+    {
+        // Ensure that this can only be run when logged in as admin
+        final String remoteUser = request.getRemoteUser();
+        if (!"admin".equals(remoteUser.toLowerCase(Locale.ROOT))) {
```
```suggestion
        if (remoteUser == null || !"admin".equals(remoteUser.toLowerCase(Locale.ROOT))) {
```

veronikaslc replied:
From what I understand there might be noone logged in as on timeout logout event or never logged in before, so potentially that line gives NPE.

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Add new cards:clarityPatientMapping and cards:clarityQuestion mapping nodes to under /apps/cards/clarityImport

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportConfig.java:
```diff
@@ -0,0 +1,44 @@
+import org.osgi.service.metatype.annotations.Designate;
+
+@Component(immediate = true, service = ClarityImportConfig.class)
+@Designate(ocd = ClarityImportConfigDefinition.class, factory = true)
```
Technically, this should not be a factory, since we only want one importer to be defined. Changing that would involve:

- removing this class
- removing the `name` attribute
- moving the `@Designate` annotation, minus the `factory` setting, directly on the `NightlyClarityImport` class
- changing the methods that handle the configuration change from `configAdded/configRemoved` to `activate/deactivate`
- making sure that if the schedule config is empty, nothing is actually scheduled

sdumitriu commented on modules/clarity-integration/src/main/features/feature.json:
```diff
@@ -0,0 +1,47 @@
+        "service.ranking:Integer":300,
+        "scripts": [
+          "create path (sling:Folder) /prems",
+          "create service user clarity-import \n set ACL on /Questionnaires,/SubjectTypes,/prems \n   allow jcr:read for clarity-import \n end \n set ACL on /Forms,/Subjects \n   allow jcr:read,rep:write,jcr:versionManagement for clarity-import \n end"
```
It would be simpler to replace the `set acl on path for user` syntax with `set acl for user on path`.

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1204/commits/50dbfd6c9f3ea466285f36d74442ad1752804af0 :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/pom.xml:
```diff
@@ -0,0 +1,135 @@
+    </plugins>
+  </build>
+
+  <dependencies>
```
Unused dependencies that should be removed:
- jackrabbit-spi-commons
- oak-store-spi
- org.apache.sling.scripting.sightly.runtime
- commons-csv
- oak-api
- commons-lang3

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1204/commits/d9b3f553fc6787fea793bc1bcb09b3b64ec62e59 :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/NightlyClarityImport.java:
```diff
@@ -0,0 +1,99 @@
+        final Runnable importJob;
+        importJob =
+            new ClarityImportTask(this.resolverFactory,
+                this.rrp.getThreadResourceResolver().getResource(MAPPING_CONFIG));
```
I think `getThreadResourceResolver` returns `null` in a background job.

Also, it is better to obtain a resource when it is needed, this code could in theory have a lock on this resource for a really long time, leading to a stale session, and a view of the mapping that doesn't take into account runtime changes to it.

I think the task itself must read the configuration, using its own resolver obtained from the factory.

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1204/commits/4837fef8ebfeb5c7618dad14308db24442720698 :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/src/main/resources/SLING-INF/nodetypes/mapping.cnd:
```diff
@@ -0,0 +1,44 @@
+  // Hardcode the resource supertype: the dataQuery is a resource
+  - sling:resourceSuperType (STRING) = "cards/Resource" mandatory autocreated protected
+  // Set a default title
+  - title (String) = "CARDS query" mandatory autocreated
```
These can be removed.

sdumitriu commented on modules/clarity-integration/src/main/resources/SLING-INF/nodetypes/mapping.cnd:
```diff
@@ -0,0 +1,44 @@
+<cards = 'https://cards.uhndata.io/'>
+
+//-----------------------------------------------------------------------------
+// Nodetype for data query handling
```
Wrong comment.

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1204/commits/32e9e93749cdd7cd5f7700fbb608fb89eefbb1cb :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/src/main/resources/SLING-INF/content/mapping.xml:
```diff
@@ -0,0 +1,90 @@
+<!--
```
I think this should be inside the `prems` project.

IntegralProgrammer replied:
Addressed in https://github.com/data-team-uhn/cards/pull/1204/commits/87ccf81628e3691f5bcd76748bd04a26810847e1 :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/pom.xml:
```diff
@@ -0,0 +1,135 @@
+            <Include-Resource>{maven-resources}</Include-Resource>
+            <Sling-Nodetypes>SLING-INF/nodetypes/mapping.cnd</Sling-Nodetypes>
+            <Sling-Initial-Content>
+              SLING-INF/content/mapping.xml;path:=/prems/mapping;overwrite:=true,
```
`/apps/cards/clarityImport`

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/NightlyClarityImport.java:
```diff
@@ -0,0 +1,99 @@
+
+    private static final String SCHEDULER_JOB_PREFIX = "NightlyClarityImport-";
+
+    private static final String MAPPING_CONFIG = "/prems/mapping";
```
`/apps/cards/clarityImport`

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -41,6 +41,8 @@
         // Access rights for the special "patient" user: do not allow access to PII
         "create user patient \n set ACL for patient \n     deny jcr:all on /Questionnaires restriction(rep:itemNames,time,location,provider) \n     deny jcr:read on /Forms restriction(cards:question,/Questionnaires/Visit*information/time,/Questionnaires/Visit*information/location,/Questionnaires/Visit*information/provider) \n     deny jcr:all on /Forms restriction(cards:questionnaire,/Questionnai ...
+        // Path for us to store mapping information within
+        "create path (sling:Folder) /proms",
```
`prems` or `proms`? Also, the rights are set in a different module, and the order in which the modules are loaded is not fixed, so the ACL may run before the path is created, which will lead to an exception. Move this in `modules/clarity-integration/src/main/features/feature.json`.

IntegralProgrammer replied:
`/prems` folder has been removed (https://github.com/data-team-uhn/cards/pull/1204/commits/2f8587ee095ac5a15aafe1246d3667226fce1c77).

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1204/commits/61f359c98b78d390448fbb84da4b6c8bb9955147 :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,506 @@
+    private static final String VALUE_PROP = "value";
+
+    /** Sling property name for the value field on an Answer node. */
+    private static final String STATUS_FIELD_PROP = "statusFlags";
```
Ugh, why do we have to manually set these? Shouldn't the flags be automatically computed by a commit editor?

IntegralProgrammer replied:
I don't know why this code was placed here in the first place but I removed it in https://github.com/data-team-uhn/cards/pull/1204/commits/ebf97fa35a001200656503964028b2395c6bcb30 and the behavior hasn't changed. That being said, the newly created Forms (_Visit information_ and _Patient information_) appear as `INCOMPLETE` when created as a result of running a Clarity import. I'm not sure if that's okay or if it even matters.

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,506 @@
+
+    private final Map<String, String> questionnaireToSubjectID;
+
+    private ThreadLocal<List<String>> nodesToCheckin = ThreadLocal.withInitial(LinkedList::new);
+
+    private ThreadLocal<VersionManager> versionManager = new ThreadLocal<>();
+
+    private ThreadLocal<String> previousPatientId = new ThreadLocal<>();
+
+    private ThreadLocal<Resource> previousPatientResource = new ThreadLocal<>();
+
+    private ThreadLocal<Boolean> createdPatientInformation = new ThreadLocal<>();
```
All these threadlocals should be cleaned up when the import is done, otherwise they can leak memory, or spill state into the next scheduled import.

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1204/commits/3fc09859ee497293b750ac545e7dc7ccbf40e4ad :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,506 @@
+            Resource formsHomepage = resolver.resolve("/Forms");
+            while (results.next()) {
+                LOGGER.info("Entry found: " + results.getString("PAT_MRN"));
+                Resource subjectParent = resolver.resolve("/Subjects");
```
This could be declared outside the loop, since it's the same for all rows.

IntegralProgrammer replied:
`subjectParent` is not the same for all rows iterated through by `results.next()` as it gets updated for each iteration of the `                for (Map.Entry<String, List<QuestionInformation>> entry : this.questionnaireToQuestions.entrySet())` loop with the call to `subjectParent = createNodeFromEntry(resolver, results, entry.getValue(), questionnaire, subjectParent, formsHomepage);`.

This, however, unveils a much larger problem as we are assuming that `this.questionnaireToQuestions.entrySet()` returns an **ordered** list of _Questionnaires_ in such an order where each _Questionnaire_ has a _Subject_ that is a child subject of the previous _Questionnaire_ in the _ordered_ list. For example, `this.questionnaireToQuestions.entrySet()` should return `["/Questionnaires/Patient information", "/Questionnaires/Visit information"]` as the _Subject_ of `/Questionnaires/Patient information` (`/SubjectTypes/Patient`) is the _parent_ _Subject_ of `/Questionnaires/Visit information`(`/SubjectTypes/Patient/Visit`).

So there are two issues that need to be resolved:

1. Can we assume that each Clarity SQL row will only generate/update a hierarchy of Subjects where each of those Subjects has no more than _one_ child Subject?
2. What's the best way to encode this information in the JCR? Currently, we are using the JCR nodes under `/apps/cards/clarityImport/` to provide a mapping of _SQL Columns_ to _CARDS Questions_ - with the nodes that provide these mappings named as `00000001`, `00000002`, ... , `0000000N`. We could modify our Java code to ensure that we iterate through these nodes in proper numerical order and we would make sure that we list these `cards:clarityQuestionMapping` nodes in the order of Subject dependency - ie. `cards:clarityQuestionMapping` nodes that refer to _Questions_ from `/Questionnaires/Patient information` would be listed before _Questions_ from `/Questionnaires/Visit information`.

How should we handle this?

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,506 @@
+        SQLException
+    {
+        String subjectID = this.questionnaireToSubjectID.get(questionnairePath);
+        subjectID = "".equals(subjectID) ? UUID.randomUUID().toString() : result.getString(subjectID);
```
The same variable is used to hold two different kinds of information: the name of the column holding the subject identifier, and the actual identifier found in a row. That's not good, took me a while to figure out what's going on here.

IntegralProgrammer replied:
Addressed in https://github.com/data-team-uhn/cards/pull/1204/commits/fd4b6b875f312dfed50fddb55119673bd6e10aa5 :heavy_check_mark:.

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,506 @@
+        final ResourceResolver resolver, final Resource parent) throws RepositoryException, PersistenceException
+    {
+        String subjectMatchQuery = String.format(
+            "SELECT * FROM [cards:Subject] as subject WHERE subject.'identifier'='%s'", identifier);
```
This works in the prems project, where subjects are always the same, and visits never change. In proms mode, where we may find the same visit multiple times in query results, it may mix up visits with similar identifiers for different patients. For future proofing, I suggest adapting the way `DataImportServlet` looks for subjects.

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Allow jcr:read access to the /apps node (and descendants) for the clarity-import service user

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Move MAPPING_CONFIG from /prems/mapping to /apps/cards/clarityImport

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Populate this.columns, this.questionnaireToQuestions, this.questionnaireToSubjectType, and this.questionnaireToSubjectID from the configuration specified in the JCR under /apps/cards/clarityImport

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup, remove debug logging

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup, remove useless /prems folder in the JCR

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup, remove unused Maven dependencies

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Code cleanup, fixed comment in modules/clarity-integration/src/main/resources/SLING-INF/nodetypes/mapping.cnd

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Move clarityMapping.xml from the clarity-integration module into prems-resources

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactor code, use better naming for variables

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

generate_compose_yaml.py: Default to exposing the MS SQL container to 127.0.0.1:1433 if --expose_mssql is specified without any port number argument

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Remove useless /proms folder

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Use same remoteUser == null check as is being used in the Heracles ExportEndpoint and in the email-notifications EmailTestEndpoint.

Co-authored-by: veronikaslc <veronika.koltunov@gmail.com>

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Cleanup all ThreadLocals at the end of the run() method

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Read the /apps/cards/clarityImport configuration during the execution of the NightlyClarityImport task as opposed to during the scheduling of that task

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

No need to manually set the statusFlags field for the created Forms

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Simplify the syntax for configuring the ACLs for the clarity-import user

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Restructure how the Clarity-to-CARDS mapping is configured under /apps/cards/clarityImport so that Subjects can be created in the correct order (ie. parent subjects before child subjects)

TODO: Modify the backend Java files to support this change

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Modified ClarityImportTask.java to support new structure of Clarity-to-CARDS mapping configuration JCR data

TODO: Get this to work with branched hierarchies and not just simple (single child) hierarchies

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Re-wrote large parts of ClarityImportTask.java to improve code robustness and support the new format of Clarity import configuration at the /apps/cards/clarityImport JCR node

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Remove unnecessary SuppressWarnings directive (checkstyle:MultipleStringLiterals) for the processClarityToCardsMapping method

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Remove unnecessary debug logging

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Move creation of jcr:Form node to outside of the try..catch block as the call to createFormsAndSubjects is, itself, within a try..catch block

Re-introduced SuppressWarnings("checkstyle:MultipleStringLiterals") as the compiler was throwing an error once again

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Remove need for warning suppression of checkstyle:MultipleStringLiterals by creating the QUESTION_PROP constant

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Remove warning suppression for checkstyle:NPathComplexity

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Remove warning suppression for checkstyle:JavaNCSS

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Removed more unnecessary debug logging

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Removed useless call to resolver.resolve("/Subjects")

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Use this.versionManager.get() instead of formNode.adaptTo(Node.class).getSession().getWorkspace().getVersionManager() for obtaining a version manager

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Use String constants for repeated String literals

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Use more meaningful local variable names in the walkThroughClarityImport method

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactor by creating the generateAnswerNodeProperties(ResourceResolver resolver, Resource questionMapping, ResultSet sqlRow) method

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactor code using a ternary expression instead of an if..else statement

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Refactor code by adding the updateExistingForm and populateEmptyForm methods

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Rework code and make it more efficient by storing the Clarity Import configuration tree as a set of Java objects within the Clarity Import task at the beginning of that task so that JCR accesses to /apps/cards/clarityImport need not take place upon the handling of each returned Clarity SQL result row.

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

In the CARDS4PREMS Clarity Import configuration add support for importing the patient first name (PAT_FIRST_NAME) and patient last name (PAT_LAST_NAME)

Also add support for generating valid PAT_FIRST_NAME and PAT_LAST_NAME test cases to generate_test_sql.py

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Removed debug logging code

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Switch from using an OSGi configuration factory for configuring the Clarity Importer to a singular OSGi configuration as there is only (at most) one Clarity Importer that needs to be configured.

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Use LOGGER.info instead of LOGGER.warn for displaying Clarity Importer activation message

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Use LOGGER.info instead of LOGGER.debug for displaying Clarity Importer deactivation message

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Rename SCHEDULER_JOB_PREFIX constant to SCHEDULER_JOB_NAME

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

LOGGER.info() whenever the ClarityImportTask runs

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Do not schedule anything if the schedule config is empty

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Today's run of the ClarityImportTask should only read Clarity SQL entries from yesterday (and not today) so as to be sure not to miss any SQL entries that were loaded today but after the execution of the ClarityImportTask

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Removed trailing whitespaces

IntegralProgrammer commented:
Rebuilt many large parts of this PR to address https://github.com/data-team-uhn/cards/pull/1204#discussion_r1046496029 and have rebased it off of the latest `dev` branch. Tested with the included `importTestSql.sql` file as well as a generated SQL file created with the `generate_test_sql.py` program and seems to work fine with the exception that all created Forms are marked as `INCOMPLETE` (_draft_) as previously mentioned in https://github.com/data-team-uhn/cards/pull/1204#discussion_r1046366117. I'm not sure whether or not that's a blocker for the PREMs use-case.

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clarityMapping.xml:
```diff
@@ -0,0 +1,169 @@
+          <node>
+            <name>Visit information</name>
+            <primaryNodeType>cards:clarityQuestionnaireMapping</primaryNodeType>
+            <node>
+              <name>00000001</name>
+              <primaryNodeType>cards:clarityQuestionMapping</primaryNodeType>
+              <property>
+                <name>sqlColumn</name>
+                <value>DISCH_DEPT_NAME</value>
+                <type>String</type>
+              </property>
+              <property>
+                <name>question</name>
+                <value>/Questionnaires/Visit information/clinic</value>
+                <type>String</type>
+              </property>
+            </node>
```
Because of the way we had to quickly adapt the PROMs code to the PREMs use cases, some of the field mapping will be counter-intuitive.
* `/Questionnaires/Visit information/clinic` needs to be populated with the patient cohort id, where a cohort is the subset of patients that receive a certain survey (such as long form survey, emergency dept survey, rehab survey, etc; this field could be interpreted as a "clinic type", e.g. they went to the emergency dept, doesn't matter at which hospital). This is under the scope of [CARDS-1963](https://phenotips.atlassian.net/browse/CARDS-1963) (which is blocked by [CARDS-1968](https://phenotips.atlassian.net/browse/CARDS-1968)) and will be computed dynamically based on other fields from the clarity dump. You can leave `/Questionnaires/Visit information/clinic` empty for now until 1963 is implemented.
* `DISCH_DEPT_NAME` needs to be placed in the ` /Questionnaires/Visit information/provider`, as the surveys technically evaluate the service of all the providers in that department.
```suggestion
            <node>
              <name>00000001</name>
              <primaryNodeType>cards:clarityQuestionMapping</primaryNodeType>
              <property>
                <name>sqlColumn</name>
                <value>DISCH_DEPT_NAME</value>
                <type>String</type>
              </property>
              <property>
                <name>question</name>
                <value>/Questionnaires/Visit information/provider</value>
                <type>String</type>
              </property>
            </node>
```
 Please note that the `provider` field in `Visit information` is multi-valued, hopefully the current implementation won't break because of this.

IntegralProgrammer replied:
The current implementation does not _break_ because of this but what happens is the `value` property of the Answer node is single-valued even though its corresponding `cards:Question` node specifies the property to be multi-valued. Is this okay or should the `ClarityImportTask` be modified so that when a single value is imported as an answer to a multi-valued question, the answer is imported as a list of _length = 1_ as opposed to just a single value?

marta- replied:
Tbh I'm not sure what will happen downstream if this is stored as single value. @sdumitriu ?

marta- replied:
I think we should just make sure it's stored exactly as it would be if the value had been entered in the form.

IntegralProgrammer replied:
Okay, I agree, it would be best to make sure that the Clarity Importer creates _answer_ nodes in the exact same way that a user manually creating forms would. I will implement that.

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1204/commits/60d2dd67db19042eb29e6a210d300b5d2f6d9172 :heavy_check_mark:.

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clarityMapping.xml:
```diff
@@ -0,0 +1,169 @@
+              <primaryNodeType>cards:clarityQuestionMapping</primaryNodeType>
+              <property>
+                <name>sqlColumn</name>
+                <value>ENTRY_TIME</value>
```
In `/Questionnaires/Visit information/time` we need to record the discharge time. I believe that's `HOSP_DISCHARGE_DTTM`.

IntegralProgrammer replied:
So would that be _instead_ of the `ENTRY_TIME` Clarity SQL column - ie. we need to map the SQL column `HOSP_DISCHARGE_DTTM` to `/Questionnaires/Visit information/time` and not do anything with the `ENTRY_TIME` SQL column?

marta- replied:
that is correct ☝️

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1204/commits/0b794bca6163358b13c062935b87d01e71b9771c :heavy_check_mark:.

marta- commented on prems-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clarityMapping.xml:
```diff
@@ -0,0 +1,169 @@
+          <name>questionnaires</name>
+          <primaryNodeType>sling:Folder</primaryNodeType>
+          <node>
+            <name>Visit information</name>
```
We also need to store `DISCH_LOC_NAME` in `/Questionnaires/Visit information/location`.

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1204/commits/4a4aa374e36f80f8590e852c27137f294dae7bbf :heavy_check_mark:.

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Change mapping of SQL column DISCH_DEPT_NAME from /Questionnaires/Visit information/clinic to /Questionnaires/Visit information/provider.

Co-authored-by: Marta Girdea <marta@phenotips.org>

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Add a CARDS4PREMS Clarity mapping from DISCH_LOC_NAME to /Questionnaires/Visit information/location

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Change the CARDS4PREMS Clarity mapping of /Questionnaires/Visit information/time from ENTRY_TIME to HOSP_DISCHARGE_DTTM

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Update importTestSql.sql to support revised Clarity SQL columns

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Update generate_test_sql.py to support revised Clarity SQL columns

IntegralProgrammer committed:
CARDS-1780: Configure/Implement data import from Clarity

Fix ClarityImportTask.java so that if a Questionnaire specifies an answer to be multi-valued, the import task creates the answer value property as a multi-value property containing only one value

sdumitriu commented on prems-resources/feature/src/main/features/feature.json:
```diff
@@ -28,6 +28,10 @@
       "id":"${project.groupId}:cards-visit-subject-type:${project.version}",
```
I think the changes in this file can be reverted; having just the jar declared as a dependency here will not work, since the feature file will not be included, and the user mapping is already present in the other feature file.

sdumitriu commented on modules/clarity-integration/src/main/java/io/uhndata/cards/clarity/importer/ClarityImportTask.java:
```diff
@@ -0,0 +1,562 @@
+
+        // Connect via SQL to the server
+        try (Connection connection = DriverManager.getConnection(connectionUrl);
+            ResourceResolver resolver = this.resolverFactory.getServiceResourceResolver(null)) {
```
You must do something like [this](https://github.com/data-team-uhn/cards/blob/eea1cfa7b71f049b0b164d9f3a3bd2b3288a0cff/heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/PauseResumeFormEditor.java#L129-L146) to make the resource resolver available to other code.

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Reorganize the methods in a top-to-bottom order

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Refactoring: made class fieds final

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Push the created resolver factory into the provider to be usable later (for CARDS-1887)

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Revert unneeded changes to the PREMS feature file

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Add support for ignored columns (only obained from Clarity, but not persisted in CARDS)
and computed columns (not obtained from Clarity, but computed and persisted in CARDS)

sdumitriu commented on modules/clarity-integration/src/main/resources/SLING-INF/nodetypes/mapping.cnd:
```diff
@@ -0,0 +1,69 @@
+  // Attributes:
+  query
+
+  // Path to a question in the parent questionnaire
+  - question (STRING) mandatory
+  // The name of the column; this is usually an sql column in Clarity, except for computed questions
+  - column (STRING) mandatory
+  // If true, the answer isn't coming from Clarity, but is computed by a data processor
+  - computed (BOOLEAN) = false mandatory autocreated
```
Changed behavior:
- question = X, column = Y, computed = false -> the column Y is queried from Clarity and stored in the target question X
- question = X, column = Y, computed = true -> the column Y is NOT queried from Clarity, but is expected to be computed by a processor, and is then stored in the target question X
- question = empty, column = Y, computed = false -> the column Y is queried from Clarity, is usable in data processors, but is not stored in CARDS
- question = empty, column = Y, computed = true -> the column Y is NOT queried from Clarity, but is computable and usable in data processors, and is not stored in CARDS; this is just temporary data, so it should not be specified in the mapping at all
- the column name cannot be empty

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/FilterClinicFrequency.java:
```diff
@@ -0,0 +1,90 @@
+ */
+@Component()
+@Designate(ocd = FilterClinicFrequency.FilterClinicFrequencyConfigDefinition.class)
+public class FilterClinicFrequency implements ClarityDataProcessor
```
This is work in progress, mostly for my testing. Things to decide:

- is the configuration as a single vector `[(department name = frequency)...]` OK? The alternative is to have multple instances of this filter active, one for each department
- if a department is not configured, all data from the department is allowed; should this be reversed, no config -> discard all?
- is the `DISCH_DEPT_NAME` the correct column to filter by?

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/FilterEmailConsent.java:
```diff
@@ -0,0 +1,91 @@
+ */
+@Component()
+@Designate(ocd = FilterEmailConsent.FilterEmailConsentConfigDefinition.class)
+public class FilterEmailConsent implements ClarityDataProcessor
```
This is work in progress, mostly for my testing. Things to decide:

- this runs first, if we cannot send emails to the patient, we shouldn't import this at all; is this OK?
- the "department that want to ignore consent" configuration is probably not what we want, but I included it temporarily for me to test things; do we actually want to keep this in production?

marta- commented:
@sdumitriu This branch is getting out of control. IIRC according to our best practices, the changes related to 1968&co should have been committed in a separate branch dependent on this one and published in a separate DRAFT PR.

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Added toString methods for easier debugging of mappings

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Bugfix: The Clinic answer is created with the wrong type

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Add support for loading data from a day other than yesterday through the import endpoint

marta- reviewed (approved):
The logic and current configuration look right to me. Didn't do any testing.

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
More descriptive name for the mapping cnd file

sdumitriu committed:
CARDS-1780: Configure/Implement data import from Clarity
Cleanup imports
